src/view/placeholder.js

```diff
@@ -145,3 +159,31 @@ function updateSinglePlaceholder( writer, element, info ) {
 
 	return changed;
 }
+
+/**
+ * Returns a view element the placeholder can be added to inside a view editing root.
+ *
+ * Even if empty, the editing root usually hosts at least one empty element (paragraph, heading, etc.).
+ * Because of that, the placeholder cannot be added directly to the root and doing so would mean both
+ * a CSS pseudo–element (with a placeholder text) and an empty element are displayed next to each other.
+ *
+ * Instead, the placeholder must be added to that empty element and this helper returns it, if there
+ * is one.
+ *
+ * @param {module:engine/view/rooteditableelement~RootEditableElement} root The root editable view that
+ * is to have a placeholder.
+ * @returns {module:engine/view/element~Element|null} An element the placeholder can be added to.
+ */
+export function getRootPlaceholderElement( root ) {
+	return () => {
```

Comment:
It is a little too magic for me. `get...Element` function does not return an element, but a function which returns element. In fact, it also decides if the placeholder should be added based on the child could, so can be used instead of `checkFunction`. Even if it let us integrate placeholders with a few lines of code, the time one will need to spent to understand how it works is not worth it. I think this logic should be some outside in the `EditorUI` class.

Reply:
I think such code in the editor UI would be more readable:
```js
let elementWithPlaceholder;
const viewRoot = view.document.getRoot();

model.document.on( 'change', () => {
	const firstRootChild = viewRoot.getChild( 1 );

	// Remove placeholder when is not needed.
	if ( viewRoot.childCount != 1 || !firstRootChild || firstRootChild.is( 'element' ) ) {
		if ( elementWithPlaceholder ) {
			removePlaceholder( view, elementWithPlaceholder );
			elementWithPlaceholder = undefined;
		}

		return;
	}

	// Attach placeholder only if firstRootChild element changed.
	if ( elementWithPlaceholder !== firstRootChild ) {
		addPlaceholder( view, firstRootChild, label );
		elementWithPlaceholder = firstRootChild;
	}
}, { priority: 'low' } );
```

Then it can be change to a utils.

Reply:
I have mixed feelings about this piece of code you proposed. Where would you like to keep it?